ui/phase info fix

vallq opened the pull request:
## What's new
continuing from #422 

Option 1:
https://user-images.githubusercontent.com/28668944/128492109-d735c671-caac-41db-a99a-706a4a005414.mp4
I tried my best with this but the expansion of each entry will always open the side panel

Option 2 (as Yadu mentioned about having the phase info in the side panel):
![image](https://user-images.githubusercontent.com/28668944/128493120-b0a3f583-b149-4697-9be7-2f63f1f0ce38.png)
![timeline](https://user-images.githubusercontent.com/28668944/128492980-6958173d-5704-4643-93b3-e2e8f754b151.png)

## Self-checks

- [x] I'm familiar with and follow this [Typescript guideline](https://basarat.gitbook.io/typescript/styleguide)
- [x] I added unit-tests for new components
- [ ] I tried testing edge cases
- [ ] I tested the behavior of the components that interact with the backend, with an e2e test

## Discussion

vallq commented:
Hi @koonpeng and @Yadunund ,
I added 2 panels to the task page in the dashboard to show the different options we have, and I will proceed to make amendments in the modified files once we've decided on which to go with / other options. Let me know what you both think! If need be, we could poll on the sg channel as well.

koonpeng commented:
I haven't tested it yet, judging from the video, I can see the problems with option 1.

* When you hover on top of the row, it is highlighted but the cursor doesn't change to a pointer, the highlight suggests it is clickable but the pointer does suggests otherwise.
  * If you make the pointer change, then it is extra confusing as there is nothing that indicates clicking on the down arrow has a different behavior.
* Clicking on the task only shows it on the side panel, with no indication that the down arrow is also clickable to show the phases.
  * This is in contrast to a typical accordion where the whole header is clickable.

koonpeng commented on packages/react-components/lib/tasks/task-timeline.tsx:
```diff
@@ -0,0 +1,122 @@
+              </Typography>
+            </TimelineOppositeContent>
+            <TimelineSeparator>
+              <TimelineDot className={timelineDotProps[idx].className} />
```
since the object is called `timelineDotProps`, I would expect it to be passed directly to `TimelineDot`.

```suggestion
              <TimelineDot {...timelineDotProps[idx]} />
```

koonpeng commented on packages/react-components/lib/tasks/task-timeline.tsx:
```diff
@@ -0,0 +1,122 @@
+            <TimelineSeparator>
+              <TimelineDot className={timelineDotProps[idx].className} />
+              {idx < timelineInfo.length - 1 && <TimelineConnector />}
+              {idx == timelineInfo.length - 1 && (
+                <TimelineConnector style={{ background: 'transparent' }} />
+              )}
```
Why not just skip rendering the last connector?

vallq replied:
I think when I was implementing this, I had a situation where removing the last connector will shift the last dot and info component towards the left if the connector is not rendered, but that seems to not happen now, so I will remove this additional connector. might have missed it previously

koonpeng replied:
:+1: feel free to add this in if that problem comes back, maybe with a comment explaining why. Just want to make sure this is intended and not an oversight.

koonpeng commented on packages/react-components/lib/tasks/task-timeline.spec.tsx:
```diff
@@ -0,0 +1,34 @@
+  it('shows all task phases', () => {
+    const task = makeTask('task_0', 3, 3);
+    const root = render(<TaskTimeline taskSummary={task} />);
+    expect(root.getByText('Phase 1 test phase 1')).toBeTruthy();
```
The test is "shows all task phases", but it only checks the first task phase.

koonpeng commented on packages/react-components/lib/tasks/task-timeline.spec.tsx:
```diff
@@ -0,0 +1,34 @@
+import { TaskTimeline } from './task-timeline';
+import { makeTask } from './test-data.spec';
+
+describe('Task Timeline', () => {
```
Should add a test that it shows the time as well.

koonpeng commented on packages/dashboard/src/components/tasks/task-page.tsx:
```diff
@@ -92,23 +92,45 @@ export function TaskPage() {
-      onAutoRefresh={autoRefreshDispatcher.setEnabled}
-    />
+    <div>
+      <TaskPanel
```
I assume the 2 task panels is just for demo?

vallq replied:
yes this is just for demo purposes, will remove when we've made a decision on which alternative to use

vallq commented on packages/react-components/lib/tasks/task-table.tsx:
```diff
@@ -15,26 +20,34 @@ const useStyles = makeStyles((theme) => ({
     background: theme.palette.action.hover,
   },
   infoRow: {
+    boxShadow: `${theme.shadows[1]}`,
     '& > *': {
       borderBottom: 'unset',
     },
```
oh this was something mentioned awhile back when we were on the throwaway/demo branch that the default line was rather faint hence I thought to implement a shadow instead. I could remove this first and add it back if needed

vallq commented on packages/react-components/lib/tasks/task-table.tsx:
```diff
@@ -15,26 +20,34 @@ const useStyles = makeStyles((theme) => ({
     },
   },
   phasesCell: {
-    padding: `0 ${theme.spacing(1)}px`,
+    padding: `0 ${theme.spacing(1)}px 0 ${theme.spacing(1)}px`,
+    boxShadow: `${theme.shadows[1]}`,
+    '&:last-child': {
+      paddingRight: `${theme.spacing(1)}px`,
+    },
   },
   phasesRow: {
     marginBottom: theme.spacing(1),
+    marginTop: theme.spacing(1),
   },
 }));
```
yea! but I was thinking we could have a separate story / component using the phases and keep it as an option within react-components

koonpeng replied:
The phases is in its own component, I don't think it is worth it to create different variants of the table, not unless we actually have a use case for different variants, better to keep things simple to avoid bloating it.

vallq commented on packages/react-components/lib/tasks/task-timeline.tsx:
```diff
@@ -0,0 +1,122 @@
+import React from 'react';
```
will look into this!

vallq commented:
1. removed unused code
2. fixed the info div to have a fixed width and max height, and to be scrollable when the content gets long
3. shifted the timeline to be left-aligned
4. fixed failing CI test by checking for all possible time values with regex

![image](https://user-images.githubusercontent.com/28668944/129519111-b6258683-3ed2-49ed-a722-a22887e8265a.png)

vallq commented:
@Yadunund how does this look? I think a highlight of the cell would be more consistent because I think that's what Chawin has implemented for the robots page as well, in his theming PR
![task-page-updates](https://user-images.githubusercontent.com/28668944/131600433-213d4eb6-e102-468c-90a3-11174da16184.png)

Yadunund commented:
@vallq thanks for iterating! Yes, let's have a vote in `sg-web`
